from database import (
    create_todo,
    fetch_all_todos,
    fetch_one_todo,
    remove_todo,
    update_todo,
)
from fastapi import FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from model import Todo

app = FastAPI()

origins = ["https://localhost:3000"]

app.add_middleware(
    CORSMiddleware, allow_origins=origins, allow_credentials=True, allow_methods=["*"], allow_headers=["*"]
)


@app.get("/")
def root():
    return {"message": "Root route"}


@app.get("/api/todo")
async def get_todo():
    response = await fetch_all_todos()
    return response


@app.get("/api/todo/{title}", response_model=Todo)
async def get_todo_by_title(title):
    response = await fetch_one_todo(title)
    if response:
        return response
    raise HTTPException(404, f"There is no todo item with this title {title}")


@app.post("/api/todo/", response_model=Todo)
async def add_todo(todo: Todo):
    response = await create_todo(todo.dict())
    if response:
        return response
    raise HTTPException(400, "Something went wrong / Bad request")


@app.put("/api/todo/{title}", response_model=Todo)
async def update_todo(title: str, description: str):
    response = await update_todo(title, description)
    if response:
        return response
    raise HTTPException(400, "Something went wrong / Bad request")


@app.delete("/api/todo/{title}")
async def delete_todo(title):
    response = await remove_todo(title)
    if response:
        return {"message": "Successfully deleted!"}
    raise HTTPException(400, "Something went wrong / Bad request")
